lib/src/library_scanner.dart

```diff
@@ -81,12 +81,22 @@ class LibraryScanner {
 
     var contextLocator = new ContextLocator();
     var roots = contextLocator.locateRoots(includedPaths: [packagePath]);
-    if (roots.length != 1) {
-      throw new StateError('Expected exactly one context root, got $roots');
+    var root = roots.firstWhere(
+        (r) => r.packagesFile.parent.path == packagePath,
+        orElse: () => null);
+    if (root == null) {
+      log.warning(
+          'No context root on the default path, selecting the one with the most files.');
```

Comment:
This seems a little magic to me. Does the user get this warning on the package page? Does this happen in practice for packages (which? Is that what we're finding out here? The package name, package path, and other information isn't mentioned in the log message). Is the "most files" fallback needed? It may produce the wrong result, so it could be better to have packages strictly specify the right path?

Reply:
This is for us to get notified through the logs. It is a non-blocking situation at this point, because we may have valid context roots to work with, but it does need investigation from our side.